src/controller.c

```diff
@@ -127,6 +130,94 @@ struct rspamd_controller_session {
 	gboolean is_spam;
 };
 
+static gboolean
+rspamd_is_encrypted_password (const gchar * password)
+{
+        const gchar *start, *end;
+        gint64 id;
+        gsize size;
+        gboolean ret = TRUE;
+        
+        if (g_str_has_prefix (password, "$")) {
+                /* Parse id */
+                start = password + 1;
+                end = start;
+                size = 0;
+                while (g_ascii_isdigit (*end)) { 
+                        size++;
+                        end++;
+                }
+                if (size) {
+                        id = g_ascii_strtoll (g_strndup (start, size), (char **) NULL, 10);
+                        if (!id) {
+                                ret = FALSE;
+                        }
+                } else {
+                       ret = FALSE;
+                }
+        } else {
+                ret = FALSE;
+        }
+                
+        return ret;
+}
+
+static const gchar *
+rspamd_encrypted_password_get_str (const gchar * password, gsize skip, gsize * length)
+{
+        const gchar *str, *start, *end;
+        gsize size;
+        
+        start = password + skip;   
+        end = start;
+        size = 0;
+        while (g_ascii_isdigit (*end) || g_ascii_isalpha (*end)) { 
+             size++;
+             end++;
+        }
+        if (size) {
+                str = g_strndup (start, size);
```

Comment:
You don't need to duplicate string here, just return the position found and the size.